openstack/standalone-cinder/mapper.go

```diff
@@ -0,0 +1,213 @@
+	availability := "nova"
+	// Apply ProvisionerParameters (case-insensitive). We leave validation of
+	// the values to the cloud provider.
+	for k, v := range options.Parameters {
+		switch strings.ToLower(k) {
+		case "type":
+			volType = v
+		case "availability":
+			availability = v
+		default:
+			return "", fmt.Errorf("invalid option %q", k)
+		}
+	}
+
+	opts := volumes_v2.CreateOpts{
+		Name:             name,
+		Size:             sizeGB,
+		VolumeType:       volType,
+		AvailabilityZone: availability,
+	}
+
+	vol, err := volumes_v2.Create(p.volumeService, &opts).Extract()
+
+	if err != nil {
+		glog.Errorf("Failed to create a %d GB volume: %v", sizeGB, err)
+		return "", err
+	}
+
+	glog.Infof("Created volume %v in Availability Zone: %v", vol.ID, vol.AvailabilityZone)
+	return vol.ID, nil
+}
+
+func waitForAvailableCinderVolume(p *cinderProvisioner, volumeID string) error {
+	// TODO: Implement proper polling instead of brain-dead timers
+	c := make(chan error)
+	go time.AfterFunc(5*time.Second, func() {
+		c <- nil
+	})
+	return <-c
+}
+
+func reserveCinderVolume(p *cinderProvisioner, volumeID string) error {
+	return volumeactions.Reserve(p.volumeService, volumeID).ExtractErr()
+}
+
+func connectCinderVolume(p *cinderProvisioner, volumeID string) (volumeConnection, error) {
+	opt := volumeactions.InitializeConnectionOpts{
+		Host:      "localhost",
+		IP:        "127.0.0.1",
+		Initiator: initiatorName,
+	}
+	var rcv rcvVolumeConnection
+	err := volumeactions.InitializeConnection(p.volumeService, volumeID, &opt).ExtractInto(&rcv)
+	if err != nil {
+		glog.Errorf("failed to initialize connection :%v", err)
+		return volumeConnection{}, err
+	}
+	glog.Infof("Received connection info: %v", rcv)
+	return rcv.ConnectionInfo, nil
+}
+
+func disconnectCinderVolume(p *cinderProvisioner, volumeID string) error {
+	opt := volumeactions.TerminateConnectionOpts{
+		Host:      "localhost",
+		IP:        "127.0.0.1",
+		Initiator: initiatorName,
+	}
+
+	err := volumeactions.TerminateConnection(p.volumeService, volumeID, &opt).Result.Err
+	if err != nil {
+		glog.Errorf("Failed to terminate connection to volume %s: %v",
+			volumeID, err)
+		return err
+	}
+
+	return nil
+}
+
+func unreserveCinderVolume(p *cinderProvisioner, volumeID string) error {
+	return volumeactions.Unreserve(p.volumeService, volumeID).ExtractErr()
+}
+
+func deleteCinderVolume(p *cinderProvisioner, volumeID string) error {
+	err := volumes_v2.Delete(p.volumeService, volumeID).ExtractErr()
+	if err != nil {
+		glog.Errorf("Cannot delete volume %s: %v", volumeID, err)
+	}
+
+	return err
+}
+
+type volumeMapper interface {
+	BuildPVSource(ctx provisionCtx) (*v1.PersistentVolumeSource, error)
+	AuthSetup(ctx provisionCtx) error
+	AuthTeardown(ctx deleteCtx) error
+}
+
+func newVolumeMapperFromConnection(conn volumeConnection) (volumeMapper, error) {
+	switch conn.DriverVolumeType {
+	default:
+		msg := fmt.Sprintf("Unsupported volume type: %s", conn.DriverVolumeType)
```

Comment:
nit: persistent volume type (and below too)

Reply:
done